runtime/protocol/aws-json-protocols/common/src/aws/smithy/kotlin/runtime/awsprotocol/json/RestJsonErrorDeserializer.kt

```diff
@@ -47,19 +47,14 @@ public object RestJsonErrorDeserializer {
 
     public fun deserialize(headers: Headers, payload: ByteArray?): ErrorDetails {
         var code: String? = headers[X_AMZN_ERROR_TYPE_HEADER_NAME]
-        var message: String? = headers[X_AMZN_ERROR_MESSAGE_HEADER_NAME]
-        if (message == null) {
-            message = headers[X_AMZN_EVENT_ERROR_MESSAGE_HEADER_NAME]
-        }
+        var message: String? = if (headers[X_AMZN_ERROR_MESSAGE_HEADER_NAME] == null) headers[X_AMZN_EVENT_ERROR_MESSAGE_HEADER_NAME] else headers[X_AMZN_ERROR_MESSAGE_HEADER_NAME]
 
         if (payload != null) {
-            val deserializer = JsonDeserializer(payload)
-            deserializer.deserializeStruct(OBJ_DESCRIPTOR) {
-                loop@while (true) {
+            JsonDeserializer(payload).deserializeStruct(OBJ_DESCRIPTOR) {
+                loop@ while (true) {
                     when (findNextFieldIndex()) {
-                        ERR_CODE_ALT1_DESCRIPTOR.index,
-                        ERR_CODE_ALT2_DESCRIPTOR.index,
-                        -> code = deserializeString()
+                        ERR_CODE_ALT1_DESCRIPTOR.index -> if (code == null) code = deserializeString()
+                        ERR_CODE_ALT2_DESCRIPTOR.index -> if (code == null) code = deserializeString()
```

Comment:
**Correctness**: I don't think this ensures the precedence order we want. If ALT2 is parsed first and ALT1 is found later, it will not override the code, but it should because it has higher precedence.